prisma-fmt/src/hover.rs

```diff
@@ -0,0 +1,194 @@
+                target_top.documentation().unwrap_or(""),
+                target_top.get_type(),
+                name,
+                relation_info,
+            ))
+        }
+        ast::SchemaPosition::Model(_, model_position) => {
+            info!("We are here {:?}", model_position);
+            None
+        }
+        psl::schema_ast::ast::SchemaPosition::CompositeType(_composite_id, composite_positon) => {
+            info!("We are here: {:?}", composite_positon);
+            None
+        }
+        psl::schema_ast::ast::SchemaPosition::Enum(_enum_id, enum_position) => {
+            info!("We are here: {:?}", enum_position);
+            None
+        }
+        psl::schema_ast::ast::SchemaPosition::DataSource(_ds_id, source_position) => {
+            info!("We are here: {:?}", source_position);
+            None
+        }
+        psl::schema_ast::ast::SchemaPosition::Generator(_gen_id, gen_position) => {
+            info!("We are here: {:?}", gen_position);
+            None
+        }
+    };
+
+    contents.map(|contents| Hover { contents, range: None })
+}
+
+fn get_relation_info<'a>(
+    ctx: &'a LSPContext<'a, HoverParams>,
+    model: &Model,
+    target_model: &'a ast::Model,
+    field: &'a Field,
+) -> Option<(String, &'a Field)> {
+    ctx.db.walk_relations().find_map(|relation| {
+        let [referencing_model, referenced_model] = relation.models();
+        let fields = relation.relation_fields().collect::<Vec<Walker<RelationFieldId>>>();
+
+        let referencing_name = ctx.db.ast(referencing_model.0)[referencing_model.1].name();
+        let referenced_name = ctx.db.ast(referenced_model.0)[referenced_model.1].name();
+
+        let self_relation = if relation.is_self_relation() { " on self" } else { " " };
+        let relation_kind = format!("{}{}", relation.relation_kind(), self_relation);
+
+        let field = if referencing_name == model.name() && referenced_name == target_model.name() {
+            if relation.is_self_relation() && fields[1].ast_field().name() == field.name() {
+                fields[0]
+            } else {
+                fields[1]
+            }
+            // * (@druue) Flipped for one-to-many relations when accessing from the block that has the many relation.
+            // ```
+            // model ModelNameA {
+            //    id  Int        @id
+            //    bId Int
+            //    val ModelNameB @relation(fields: [bId], references: [id])
+            // }
+            //
+            // model ModelNameB {
+            //   id Int          @id
+            //   A  ModelNameA[]
+            //      ^^^^^^^^^^ // When accessing from here
+            // }
+            // ```
+            // * I have no idea why.
+        } else if referencing_name == target_model.name() && referenced_name == model.name() {
+            fields[0]
+        } else {
+            warn!("couldn't find relation");
+            return None;
+        };
```

Comment:
This is the weirdness that I mentioned @Weakky

Reply:
It would help if you used walkers to abstract traversing relations 👇

```rust
        psl::schema_ast::ast::SchemaPosition::Model(
            model_id,
            ModelPosition::Field(field_id, FieldPosition::Type(name)),
        ) => {
            let initiating_field = &ctx.db.walk((ctx.initiating_file_id, model_id)).field(field_id);

            match initiating_field.refine() {
                psl::parser_database::walkers::RefinedFieldWalker::Scalar(sf) => match sf.scalar_field_type() {
                    psl::parser_database::ScalarFieldType::CompositeType(ct) => {
                        let ct = ctx.db.walk(ct);
                        // Continue here

                    },
                    _ => None,
                },
                psl::parser_database::walkers::RefinedFieldWalker::Relation(rf) => {
                    let relation = rf.relation();
                    let opposite_model = rf.related_model();
                    let opposite_field = rf.opposite_relation_field().unwrap().ast_field();
                    let self_relation = if relation.is_self_relation() { " on self" } else { " " };
                    let relation_kind = format!("{}{}", relation.relation_kind(), self_relation);

                    Some(format_hover_content(
                        opposite_model.ast_model().documentation().unwrap_or_default(),
                        "model",
                        name,
                        Some((relation_kind, opposite_field)),
                    ))
                }
            }
        }
```